src/amp-story-embed.js

```diff
@@ -123,34 +194,145 @@ export class AmpStoryEmbed {
   }
 
   /**
-   * @return {!Promise}
    * @public
    */
   layoutCallback() {
-    // TODO: Layout all child iframes.
-    return this.layoutIframe_(0);
+    if (this.isLaidOut_) {
+      return;
+    }
+
+    const iframes = toArray(this.rootEl_.querySelectorAll('iframe'));
+    // TODO(Enriqe): Layout all child iframes.
+    this.layoutIframe_(this.stories_[0], iframes[0]);
+
+    this.isLaidOut_ = true;
   }
 
   /**
-   * @param {number} index
-   * @return {!Promise}
+   * @param {!Element} story
+   * @param {!Element} iframe
    * @private
    */
-  layoutIframe_(index) {
-    const story = this.stories_[index];
-    this.iframeEl_.setAttribute('src', story.href);
-    return Promise.resolve();
+  layoutIframe_(story, iframe) {
+    const {href} = this.getEncodedUrl_(story.href);
+
+    iframe.setAttribute('src', href);
+  }
+
+  /**
+   * Gets encoded url for viewer usage.
+   * @param {string} href
+   * @return {!Location}
+   * @private
+   */
+  getEncodedUrl_(href) {
+    const {location} = this.win_;
+    const url = parseUrlWithA(this.cachedA_, location.href);
+
+    const params = dict({
+      'amp_js_v': '0.1',
+      'visibilityState': 'inactive',
+      'origin': url.origin,
+    });
+
+    let inputUrl = addParamsToUrl(href, params);
+    inputUrl = inputUrl.replace('amp_js_v=0.1', 'amp_js_v=0.1#');
+
+    return parseUrlWithA(this.cachedA_, inputUrl);
+  }
+
+  /**
+   * Sends a message to the story document to make it visible.
+   *
+   * @private
+   * @param {string} iframeId
+   */
+  displayStory_(iframeId) {
+    this.messagingFor_[iframeId].sendRequest(
+      'visibilitychange',
+      {state: 'visible'},
+      true
+    );
   }
 }
 
-self.onload = () => {
+/**
```

Comment:
Should have mentioned that in my previous review, but this file will grow a lot, can you start thinking of an architecture to avoid having one single 3,000+ lines file?
All this build/layout logic looks like a good candidate to have its own file. Should it be a class or different helper functions? That's up to you, but I think the rest of the codebase uses almost exclusively classes, except for reusable helpers that have to be imported separately.